File: src/main/java/no/fintlabs/mapping/IntegrationMetadataMappingService.java
package no.fintlabs.mapping;

import no.fintlabs.model.dtos.IntegrationMetadataDto;
import no.fintlabs.model.entities.IntegrationMetadata;
import org.springframework.stereotype.Service;

@Service
public class IntegrationMetadataMappingService {

    private final InstanceMetadataContentMappingService instanceMetadataContentMappingService;

    public IntegrationMetadataMappingService(InstanceMetadataContentMappingService instanceMetadataContentMappingService) {
        this.instanceMetadataContentMappingService = instanceMetadataContentMappingService;
    }

    public IntegrationMetadata toEntity(IntegrationMetadataDto integrationMetadataDto) {
        return IntegrationMetadata
                .builder()
                .sourceApplicationId(integrationMetadataDto.getSourceApplicationId())
                .sourceApplicationIntegrationId(integrationMetadataDto.getSourceApplicationIntegrationId())
                .sourceApplicationIntegrationUri(integrationMetadataDto.getSourceApplicationIntegrationUri())
                .integrationDisplayName(integrationMetadataDto.getIntegrationDisplayName())
                .version(integrationMetadataDto.getVersion())
                .instanceMetadata(
                        instanceMetadataContentMappingService.toEntity(integrationMetadataDto.getInstanceMetadata())
                )
                .build();
    }

    public IntegrationMetadataDto toDto(IntegrationMetadata integrationMetadata) {
        return IntegrationMetadataDto
                .builder()
                .id(integrationMetadata.getId())
                .sourceApplicationId(integrationMetadata.getSourceApplicationId())
                .sourceApplicationIntegrationId(integrationMetadata.getSourceApplicationIntegrationId())
                .sourceApplicationIntegrationUri(integrationMetadata.getSourceApplicationIntegrationUri())
                .integrationDisplayName(integrationMetadata.getIntegrationDisplayName())
                .version(integrationMetadata.getVersion())
                .instanceMetadata(
                        instanceMetadataContentMappingService.toDto(integrationMetadata.getInstanceMetadata())
                )
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/discovery/service/mapping/IntegrationMetadataMappingService.kt
package no.novari.flyt.discovery.service.mapping

import no.novari.flyt.discovery.service.model.dtos.IntegrationMetadataDto
import no.novari.flyt.discovery.service.model.entities.IntegrationMetadata
import org.springframework.stereotype.Service

@Service
class IntegrationMetadataMappingService(
    private val instanceMetadataContentMappingService: InstanceMetadataContentMappingService,
) {
    fun toEntity(integrationMetadataDto: IntegrationMetadataDto): IntegrationMetadata =
        IntegrationMetadata(
            sourceApplicationId = requireNotNull(integrationMetadataDto.sourceApplicationId),
            sourceApplicationIntegrationId = requireNotNull(integrationMetadataDto.sourceApplicationIntegrationId),
            sourceApplicationIntegrationUri = integrationMetadataDto.sourceApplicationIntegrationUri,
            integrationDisplayName = requireNotNull(integrationMetadataDto.integrationDisplayName),
            version = requireNotNull(integrationMetadataDto.version),
            instanceMetadata =
                integrationMetadataDto.instanceMetadata?.let(instanceMetadataContentMappingService::toEntity),
        )

    fun toDto(integrationMetadata: IntegrationMetadata): IntegrationMetadataDto =
        IntegrationMetadataDto(
            id = integrationMetadata.id,
            sourceApplicationId = requireNotNull(integrationMetadata.sourceApplicationId),
            sourceApplicationIntegrationId = requireNotNull(integrationMetadata.sourceApplicationIntegrationId),
            sourceApplicationIntegrationUri = integrationMetadata.sourceApplicationIntegrationUri,
            integrationDisplayName = requireNotNull(integrationMetadata.integrationDisplayName),
            version = requireNotNull(integrationMetadata.version),
            instanceMetadata =
                integrationMetadata.instanceMetadata?.let(instanceMetadataContentMappingService::toDto),
        )
}
